app/src/main/java/com/example/traders/presentation/profile/portfolio/PortfolioViewModel.kt

```diff
@@ -1,206 +1,209 @@
-//
-//        sortedList.subList(0, 4).forEach {
-//            chartData.add(
-//                PieEntry(
-//                    it.amountInUsd.toFloat(),
-//                    it.symbol
-//                )
-//            )
-//        }
-//
-//        val othersUsdVal = sortedList.subList(4, sortedList.size).sumOf { it.amountInUsd }
-//        chartData.add(PieEntry(othersUsdVal.toFloat(), "Others"))
-//
-//        return chartData
-//    }
-//
-//    companion object {
-//        private val colors =
-//            listOf(-13710223, -932849, -1618884, -13330213, -4128884, -2164, -12148, -7542017, -29539)
-//    }
-//
-//}
\ No newline at end of file
+
+import com.example.traders.presentation.BaseViewModel
+import com.example.traders.database.Crypto
+import com.example.traders.network.models.CryptoTicker
+import com.example.traders.network.repository.CryptoRepository
+import com.example.traders.utils.roundNum
+import com.github.mikephil.charting.data.PieEntry
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.async
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.launch
+import java.math.BigDecimal
+import javax.inject.Inject
+
+@HiltViewModel
+class PortfolioViewModel @Inject constructor(
+    private val repository: CryptoRepository
+) : BaseViewModel() {
+
+    private val _state = MutableStateFlow(
+        PortfolioState(
+            colors = colors
+        )
+    )
+    val state = _state.asStateFlow()
+//    val state = _state.asStateFlow().combine(repository.getLiveAllCryptoPortfolio().asFlow()) {
+//        portfolioState, liveAllCryptoPortfolio ->
+//        updateStateData()
+//    }
+
+    val live_portfolio_list = repository.getLiveAllCryptoPortfolio()
+
+    fun chartUpdated() {
+        _state.value = _state.value.copy(chartReadyForUpdate = false)
+    }
+
+    fun updatePortfolioState() {
+        if (listNotEmptyAndNotEqualsToPrevList()) {
+            _state.value = _state.value.copy(prevList = live_portfolio_list.value ?: emptyList())
+            launch {
+
+
+                collectRequiredPrices(sublistOfNewCrypto())
+                calculateTotalBalance()
+                calculateChartData(
+                )
+            }
+        } else if (listIsEmpty()) {
+            setStateListsToEmpty()
+            launch {
+                calculateChartData()
+            }
+        }
+        // if we wrote else it would mean nothing has changed and we dont need to change anything
+    }
+
+    private fun listIsEmpty(): Boolean {
+        return live_portfolio_list.value?.isEmpty() ?: false
+    }
+
+    private fun listNotEmptyAndNotEqualsToPrevList(): Boolean {
+        return live_portfolio_list.value?.isNotEmpty() ?: false && _state.value.prevList != live_portfolio_list.value
+    }
```

Comment:
Function name is too long and should be abbreviated.
For example:
`private fun listNotEmptyAndRenewed()`